src/lib.rs

```diff
@@ -509,98 +538,123 @@ impl FailPoint {
-    ///
-    /// # Panics
-    ///
-    /// Panics if an action is not formatted correctly.
-    pub fn setup() -> Self {
-        // Cleanup first, in case of previous failed/panic'ed test scenarios.
-        let scenario_guard = SCENARIO.lock().unwrap_or_else(|e| e.into_inner());
-        let mut registry = scenario_guard.registry.write().unwrap();
-        Self::cleanup(&mut registry);
-
-        let failpoints = match env::var("FAILPOINTS") {
-            Ok(s) => s,
-            Err(VarError::NotPresent) => return Self { scenario_guard },
-            Err(e) => panic!("invalid failpoints: {:?}", e),
-        };
-        for mut cfg in failpoints.trim().split(';') {
-            cfg = cfg.trim();
-            if cfg.is_empty() {
-                continue;
-            }
-            let (name, order) = partition(cfg, '=');
-            match order {
-                None => panic!("invalid failpoint: {:?}", cfg),
-                Some(order) => {
-                    if let Err(e) = set(&mut registry, name.to_owned(), order) {
-                        panic!("unable to configure failpoint \"{}\": {}", name, e);
-                    }
-                }
-            }
+impl FailPointRegistry {
+    /// Register the current thread to this failpoints registry.
+    pub fn register_current(&self) -> Result<(), String> {
+        let id = thread::current().id();
+        let ret = REGISTRY_GROUP
+            .write()
+            .unwrap()
+            .insert(id, self.registry.clone());
+
+        if ret.is_some() {
+            Err("current thread has been registered with one registry".to_owned())
+        } else {
+            Ok(())
         }
-        Self { scenario_guard }
     }
 
-    /// Tear down the fail point system.
-    ///
-    /// Clears the configuration of all fail points. Any paused fail
-    /// points will be notified before they are deactivated.
-    ///
-    /// This function should generally be called after running a test with fail points.
-    /// Calling `teardown` without previously calling `setup` results in a no-op.
-    pub fn teardown(self) {
-        drop(self)
+    /// Deregister the current thread to this failpoints registry.
+    pub fn deregister_current(&self) {
+        let id = thread::current().id();
+        REGISTRY_GROUP.write().unwrap().remove(&id);
+    }
+
+    /// Clean up registered fail points in this registry.
+    pub fn teardown(&self) {
+        let mut registry = self.registry.write().unwrap();
+        cleanup(&mut registry);
     }
+}
 
-    /// Clean all registered fail points.
-    fn cleanup(registry: &mut std::sync::RwLockWriteGuard<'a, Registry>) {
-        for p in registry.values() {
-            // wake up all pause failpoint.
-            p.set_actions("", vec![]);
+lazy_static::lazy_static! {
+    static ref REGISTRY_GROUP: RwLock<HashMap<thread::ThreadId, Arc<RwLock<Registry>>>> = Default::default();
+    static ref REGISTRY_GLOBAL: FailPointRegistry = Default::default();
```

Comment:
Better keep the old behavior in case user prefer run tests in sequential order to get a cleaner code.

Reply:
What is the old behavior you mean?

Reply:
A preconfigured global lock. Or you can just keep the `FailScenario`.